generator/translator_planet.cpp

```diff
@@ -0,0 +1,232 @@
+  m_nodeRelations(m_routingTagsProcessor),
+  m_wayRelations(m_routingTagsProcessor),
+  m_metalinesBuilder(info.GetIntermediateFileName(METALINES_FILENAME))
+{
+  auto const addrFilePath = info.GetAddressesFileName();
+  if (!addrFilePath.empty())
+    m_addrWriter.reset(new FileWriter(addrFilePath));
+
+  auto const restrictionsFilePath = info.GetIntermediateFileName(RESTRICTIONS_FILENAME);
+  if (!restrictionsFilePath.empty())
+    m_routingTagsProcessor.m_restrictionWriter.Open(restrictionsFilePath);
+
+  auto const roadAccessFilePath = info.GetIntermediateFileName(ROAD_ACCESS_FILENAME);
+  if (!roadAccessFilePath.empty())
+    m_routingTagsProcessor.m_roadAccessWriter.Open(roadAccessFilePath);
+}
+
+void TranslatorPlanet::EmitElement(OsmElement * p)
+{
+  CHECK(p, ("Tried to emit a null OsmElement"));
+
+  FeatureParams params;
+  switch(p->type)
+  {
+  case OsmElement::EntityType::Node:
+  {
+    if (p->m_tags.empty())
+      break;
+
+    if (!ParseType(p, params))
+      break;
+
+    m2::PointD const pt = MercatorBounds::FromLatLon(p->lat, p->lon);
+    EmitPoint(pt, params, osm::Id::Node(p->id));
+    break;
+  }
+  case OsmElement::EntityType::Way:
+  {
+    FeatureBuilder1 ft;
+    m2::PointD pt;
+    // Parse geometry.
+    for (uint64_t ref : p->Nodes())
+    {
+      if (!m_holder.GetNode(ref, pt.y, pt.x))
+        break;
+      ft.AddPoint(pt);
+    }
+
+    if (p->Nodes().size() != ft.GetPointsCount())
+      break;
+
+    if (ft.GetPointsCount() < 2)
+      break;
+
+    if (!ParseType(p, params))
+      break;
+
+    ft.SetOsmId(osm::Id::Way(p->id));
+    bool isCoastLine = (m_coastType != 0 && params.IsTypeExist(m_coastType));
+
+    EmitArea(ft, params, [&] (FeatureBuilder1 & ft)
+    {
+      isCoastLine = false;  // emit coastline feature only once
+      HolesProcessor processor(p->id, m_holder);
+      m_holder.ForEachRelationByWay(p->id, processor);
+      ft.SetAreaAddHoles(processor.GetHoles());
+    });
+
+    m_metalinesBuilder(*p, params);
+    EmitLine(ft, params, isCoastLine);
+    break;
+  }
+  case OsmElement::EntityType::Relation:
+  {
+    // 1. Check, if this is our processable relation. Here we process only polygon relations.
+    if (!p->HasTagValue("type", "multipolygon"))
+      break;
+
+    if (!ParseType(p, params))
+      break;
+
+    HolesRelation helper(m_holder);
+    helper.Build(p);
+
+    auto const & holesGeometry = helper.GetHoles();
+    auto & outer = helper.GetOuter();
+
+    outer.ForEachArea(true, [&] (FeatureBuilder1::PointSeq const & pts,
+                      std::vector<uint64_t> const & ids)
+    {
+      FeatureBuilder1 ft;
+
+      for (uint64_t id : ids)
+        ft.AddOsmId(osm::Id::Way(id));
+
+      for (auto const & pt : pts)
+        ft.AddPoint(pt);
+
+      ft.AddOsmId(osm::Id::Relation(p->id));
+      EmitArea(ft, params, [&holesGeometry] (FeatureBuilder1 & ft) {
+        ft.SetAreaAddHoles(holesGeometry);
+      });
+    });
+    break;
+  }
+  default:
+    break;
```

Comment:
Let's try to get rid of the default. If we do so, we'd get a warning in case of adding extra items to OsmElement::EntityType.

Reply:
++